renovate/src/Renovate/Redirect/LayoutBlocks/Compact.hs

```diff
@@ -77,11 +77,25 @@ compactLayout startAddr strat blocks0 injectedCode cfgs = do
   -- We need this so that we can re-arrange them and preserve the fallthrough
   -- behavior of blocks ending in conditional jumps (or non-jumps).
   -- traceM (show (PD.vcat (map PD.pretty (L.sortOn (basicBlockAddress . lpOrig) (F.toList blocks)))))
-  blockChunks <- groupBlocks (grouping strat) cfgs blocks0
-  let (modifiedBlockChunks, unmodifiedBlocks_) = foldMap splitChunk blockChunks
-      unmodifiedBlocks = map noFallthroughPair unmodifiedBlocks_
+  blockChunks0 <- groupBlocks (grouping strat) cfgs blocks0
+
+  -- Some layout strategies make singletons (every block in its own group).
+  -- Others group blocks by contiguous loop body or entire function.
+  --
+  -- The key observation is that we only need our fallthrough patching code (in
+  -- concretize) to add explicit fallthroughs for the *last* block in these
+  -- groups.  We can remove all of the other fallthroughs here.
+  --
+  -- We want to do this as early as possible because the number of fallthroughs
+  -- affect the ultimate sizes of blocks, and we need to be able to accurately
+  -- compute block sizes to do layout.
+  --
+  -- It is important to do this before we compute any block sizes so that things
+  -- are all internally consistent.
+  blockChunks <- removeUnneededFallthroughs blockChunks0
```

Comment:
Love this comment. Was watching for where this behavior would happen, and this made it super easy to spot.